hoodie-client/src/main/java/com/uber/hoodie/table/HoodieCopyOnWriteTable.java

```diff
@@ -575,6 +576,41 @@ protected HoodieMergeHandle getUpdateHandle(String commitTime, String fileLoc,
     return stats;
   }
 
+  @Override
+  @SuppressWarnings("unchecked")
+  public Optional<Integer> finalizeWrite(JavaSparkContext jsc, List writeStatuses) {
+    if (!config.shouldFinalizeWrite()) {
+      return Optional.empty();
+    }
+
+    List<Tuple2<String, Boolean>> results = jsc.parallelize(writeStatuses, config.getFinalizeParallelism())
+        .map(writeStatus -> {
+          Tuple2<String, HoodieWriteStat> writeStatTuple2 = (Tuple2<String, HoodieWriteStat>) writeStatus;
+          HoodieWriteStat writeStat = writeStatTuple2._2();
+          final FileSystem fs = FSUtils.getFs();
+          final Path finalPath = new Path(config.getBasePath(), writeStat.getPath());
+
+          if (writeStat.getTempPath() != null) {
```

Comment:
why would this be null?

Reply:
This would be null for update case, i.e. data files written from `HoodieMergeHandle`.

Reply:
I see.. For consistency, should we also write to tmp and move for updates as well? I am thinking this will become a quirk in code, that everyone needs to remember

Reply:
For update heavy flows, `HoodieMergeHandle` would update more than 10 thousands of files, and if we put all those rename operations in `finalizeWrite`, the `finalizeWrite` step could take minutes to finish.  Note increasing parallelism won’t help too much here because the readwrite lock in HDFS namenode. 

On the other hand, unlike `HoodieCreateHandle`, `HoodieMergeHandle` produces idempotent data file across possible spark stage retries, to amortize the cost of `rename` for update heavy flows we may explore to do the `rename` in `HoodieMergeHandle`.  However this also means that we can’t remove the null check `if (tempPath != null)`.  

I’d suggest to explore to do the `rename` in `HoodieMergeHandle` as a separate follow-up task with proper testing.

Thoughts and suggestions?

Reply:
@ovj wdyt from an operational perspective? 

>Note increasing parallelism won’t help too much here because the readwrite lock in HDFS namenode.
Can you expand on this? does HDFS NN lock the entire table tree? can't we parallelize at the partition level.
 
Understand the additional calls needed here. That's a real concern. I am okay with just doing the insert path for now. Please file a followon task.

Reply:
@n3nash as well

Reply:
Sorry for late reply; was on vacation. HoodieMergeHandle too has the same problem of corrupted file (Success followed by Failure case) and spark not retrying that stage. I think we should do the same for HoodieMergeHandle too. Not sure if we want to handle it here or fix it in follow up task.

Reply:
Okay.. here is what I suggest.. lets implement this for MergeHandle as well, with two different parameters (both off by default) to control whether this kicks in for CreateHandle / MergeHandle. you can then choose whichever way.. 

Code also remains symmetrical and maintainable.. WDYT?

Reply:
Yeah, that sounds good, and agree, we need this for HoodieMergeHandle as well.